storage/storage_dest.go

```diff
@@ -317,122 +325,141 @@ func (s *storageImageDestination) TryReusingBlobWithOptions(ctx context.Context,
 
 	return reused, info, s.queueOrCommit(*options.LayerIndex, addedLayerInfo{
 		digest:     info.Digest,
+		tocDigest:  options.TOCDigest,
 		emptyLayer: options.EmptyLayer,
 	})
 }
 
-// tryReusingBlobAsPending implements TryReusingBlobWithOptions for (digest, size or -1), filling s.uncompressedOrTocDigest and other metadata.
+// tryReusingBlobAsPending implements TryReusingBlobWithOptions for (digestBlob, size or -1), filling s.blobDiffIDs and other metadata.
 // The caller must arrange the blob to be eventually committed using s.commitLayer().
-func (s *storageImageDestination) tryReusingBlobAsPending(digest digest.Digest, size int64, options *private.TryReusingBlobOptions) (bool, private.ReusedBlob, error) {
+func (s *storageImageDestination) tryReusingBlobAsPending(digestBlob digest.Digest, size int64, options *private.TryReusingBlobOptions) (bool, private.ReusedBlob, error) {
 	// lock the entire method as it executes fairly quickly
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
 	if options.SrcRef != nil {
 		// Check if we have the layer in the underlying additional layer store.
-		aLayer, err := s.imageRef.transport.store.LookupAdditionalLayer(digest, options.SrcRef.String())
+		aLayer, err := s.imageRef.transport.store.LookupAdditionalLayer(digestBlob, options.SrcRef.String())
 		if err != nil && !errors.Is(err, storage.ErrLayerUnknown) {
-			return false, private.ReusedBlob{}, fmt.Errorf(`looking for compressed layers with digest %q and labels: %w`, digest, err)
+			return false, private.ReusedBlob{}, fmt.Errorf(`looking for compressed layers with digest %q and labels: %w`, digestBlob, err)
 		} else if err == nil {
 			// Record the uncompressed value so that we can use it to calculate layer IDs.
-			s.uncompressedOrTocDigest[digest] = aLayer.UncompressedDigest()
-			s.blobAdditionalLayer[digest] = aLayer
+			s.blobDiffIDs[digestBlob] = aLayer.UncompressedDigest()
+			s.blobAdditionalLayer[digestBlob] = aLayer
 			return true, private.ReusedBlob{
-				Digest: digest,
+				Digest: digestBlob,
 				Size:   aLayer.CompressedSize(),
 			}, nil
 		}
 	}
 
-	if digest == "" {
+	if digestBlob == "" && options.TOCDigest == "" {
 		return false, private.ReusedBlob{}, errors.New(`Can not check for a blob with unknown digest`)
 	}
-	if err := digest.Validate(); err != nil {
-		return false, private.ReusedBlob{}, fmt.Errorf("Can not check for a blob with invalid digest: %w", err)
+	if digestBlob != "" {
```

Comment:
See elsewhere… my initial guess is that making `Digest` an entirely optional value would be a much more invasive change for the codebase.

Though it is probably plausible for an input to this function, locally.

Reply:
We are returning `ReusedBlob{Digest: blobDigest}` even in the TOC match case, so it really doesn’t make sense to accept an empty `blobDigest` in any of the code paths.